src/test/java/AuthenticatedHttpServletTest.java

```diff
@@ -0,0 +1,201 @@
+import java.io.StringWriter;
+import java.security.GeneralSecurityException;
+import javax.servlet.http.Cookie;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mockito;
+
+@RunWith(JUnit4.class)
+public class AuthenticatedHttpServletTest {
+  // Mock abstract servlet with real constructor and real methods
+  private static final AuthenticationVerifier authVerifier =
+      Mockito.mock(AuthenticationVerifier.class);
+  private static final AuthenticatedHttpServlet servlet =
+      Mockito.mock(
+          AuthenticatedHttpServlet.class,
+          Mockito.withSettings()
+              .useConstructor(authVerifier)
+              .defaultAnswer(Mockito.CALLS_REAL_METHODS));
+
+  private HttpServletRequest request;
+  private HttpServletResponse response;
+  private StringWriter stringWriter;
+  private PrintWriter printWriter;
+
+  private static final Boolean AUTHENTICATION_VERIFIED = true;
+  private static final Boolean AUTHENTICATION_NOT_VERIFIED = false;
+
+  private static final String ID_TOKEN_KEY = "idToken";
+  private static final String ID_TOKEN_VALUE = "sampleId";
+  private static final String ACCESS_TOKEN_KEY = "accessToken";
+  private static final String ACCESS_TOKEN_VALUE = "sampleAccessToken";
+
+  private static final Cookie sampleIdTokenCookie = new Cookie(ID_TOKEN_KEY, ID_TOKEN_VALUE);
+  private static final Cookie sampleAccessTokenCookie =
+      new Cookie(ACCESS_TOKEN_KEY, ACCESS_TOKEN_VALUE);
+  private static final Cookie emptyIdToken = new Cookie(ID_TOKEN_KEY, "");
+  private static final Cookie emptyAccessToken = new Cookie(ACCESS_TOKEN_KEY, "");
+
+  private static final Cookie[] noCookies = new Cookie[] {};
+  private static final Cookie[] noAccessTokenCookies = new Cookie[] {sampleIdTokenCookie};
+  private static final Cookie[] noIdTokenCookies = new Cookie[] {sampleAccessTokenCookie};
+  private static final Cookie[] emptyIdTokenCookies =
+      new Cookie[] {emptyIdToken, sampleAccessTokenCookie};
+  private static final Cookie[] emptyAccessTokenCookies =
+      new Cookie[] {sampleIdTokenCookie, emptyAccessToken};
+  private static final Cookie[] validCookies =
+      new Cookie[] {sampleIdTokenCookie, sampleAccessTokenCookie};
+
+  @Before
+  public void init() throws IOException {
+    request = Mockito.mock(HttpServletRequest.class);
+    response = Mockito.mock(HttpServletResponse.class);
+
+    // Writer used in get/post requests to capture HTTP response values
+    stringWriter = new StringWriter();
+    printWriter = new PrintWriter(stringWriter);
+    when(response.getWriter()).thenReturn(printWriter);
+  }
+
+  @After
+  public void clear() {
+    // Dump contents after each test
+    stringWriter.getBuffer().setLength(0);
+  }
+
+  @Test
+  public void getRequestNoTokens() throws IOException {
```

Comment:
Short comments describing what is handled and expected might be useful here.

Reply:
I'm hesitant to add comments here, as I think they would explain the what not the why. The name of each method describes what's being tested, and the code is pretty trivial in terms of what's expected. Do you have something specific in mind?

Reply:
I agree that you don't need to document every test case. There are some cases where a test is particularly complicated or the condition being tested is not obvious, but this test case in particular is pretty trivial.

(This is a good exception to "public methods need comments".)